jax.scipy.special: manually document functions to avoid scipy import

jakevdp opened the pull request:
This lets us give more implementation-specific information, and lets us avoid a dependency on scipy. Related to #20799.

Preview rendered docstrings here: https://jax--20822.org.readthedocs.build/en/20822/jax.scipy.html#module-jax.scipy.special

Part of #21461.

hawkinsp reviewed (commented):
Same general comments on these:

* use type links where possible for `Array` and `ArrayLike`
* at least link to definition of the math in each case (Wikipedia being good), although having formulae is better where we can.
* an example would be peachy.

Doable?

jakevdp commented:
It's all doable, it's just about how much time we want to invest

jakevdp commented:
Mathematical definitions added

hawkinsp commented on jax/_src/scipy/special.py:
```diff
@@ -32,131 +31,437 @@
+    x: arraylike, real-valued.
+
+  Returns:
+    array contining values of the logit function.
```
sp. containing

jakevdp replied:
Done

hawkinsp commented on jax/_src/scipy/special.py:
```diff
@@ -32,131 +31,437 @@
 @custom_derivatives.custom_jvp
-@implements(osp_special.xlogy, module='scipy.special')
 def xlogy(x: ArrayLike, y: ArrayLike) -> Array:
+  """Compute x*log(y) with well-defined derivatives.
```
What makes the derivatives well-defined?

You say below the primal value is defined at 0, 0, but what implications does that have for the derivative?

jakevdp replied:
They're well-defined becuase the gradients are implemented in terms of `xlogy`

jakevdp replied:
Done

hawkinsp commented on jax/_src/scipy/special.py:
```diff
@@ -171,8 +476,23 @@ def _xlogy_jvp(primals, tangents):
 @custom_derivatives.custom_jvp
-@implements(osp_special.xlog1py, module='scipy.special', update_doc=False)
 def xlog1py(x: ArrayLike, y: ArrayLike) -> Array:
+  """Compute x*log(1 + y) with well-defined derivatives.
```
Same.

jakevdp replied:
Done

jakevdp committed:
DOC: add manual documentation to jax.scipy.special functions.

This lets us give more implementation-specific information, and
lets us avoid a needless dependency on scipy.
